### PR TITLE
Add Reimbursement model and add Advance to PPM

### PR DESCRIPTION
**Pull request by macrael**
## Description

Adds the reimbursement model and makes it play nice with PPM. 

## Reviewer Notes

What do you think of my State Machine methods? Any good way we could enforce their use?

## Code Review Verification Steps

* [ ] All tests pass.
* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/blob/master/docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/blob/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* [ ] Any migrations/schema changes also update the diagram in docs/schema/dp3.sqs
* [ ] There are no aXe warnings for UI.
* [ ] This works in IE.
* Any new client dependencies (Google Analytics, hosted libraries, CDNs, etc) have been:
  * [ ] Communicated to @willowbl00
  * [ ] Added to the list of [network dependencies](https://github.com/transcom/mymove#client-network-dependencies)
* [ ] Request review from a member of a different team.
* [ ] (TRIAL) Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/157712258) for this change
* [this article](tbd) explains more about the approach used.

### CODE REVIEW

File: pkg/models/reimbursement.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	// ReimbursementStatusPAID captures enum value "PAID"
+	ReimbursementStatusPAID ReimbursementStatus = "PAID"
+)
+
+// Reimbursement is money that is intedended to be paid to the servicemember
+type Reimbursement struct {
+	ID              uuid.UUID           `json:"id" db:"id"`
+	CreatedAt       time.Time           `json:"created_at" db:"created_at"`
+	UpdatedAt       time.Time           `json:"updated_at" db:"updated_at"`
+	RequestedAmount int                 `json:"requested_amount" db:"requested_amount"`
```
Comment (akostibas) [inline pkg/models/reimbursement.go]: use `unit.Cents`

File: pkg/models/reimbursement.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+func BuildDraftReimbursement(requestedAmount int, methodOfReceipt MethodOfReceipt) Reimbursement {
+	return Reimbursement{
+		Status:          ReimbursementStatusDRAFT,
+		RequestedAmount: requestedAmount,
+		MethodOfReceipt: methodOfReceipt,
+	}
+}
+
+// BuildRequestedReimbursement makes a Reimbursement in the REQUEST state, but does not save it
+func BuildRequestedReimbursement(requestedAmount int, methodOfReceipt MethodOfReceipt) Reimbursement {
```
Comment (akostibas) [inline pkg/models/reimbursement.go]: What's this for? Why would a reimbursement be created in the Requested state?

File: pkg/models/reimbursement.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		string(MethodOfReceiptDIRECTDEPOSIT),
+		string(MethodOfReceiptOTHER),
+	}
+
+	fmt.Println("WEOIFNWOIFNOWF", r)
+
+	return validate.Validate(
+		&validators.IntIsGreaterThan{Field: r.RequestedAmount, Name: "RequestedAmount", Compared: 0},
+		&validators.StringInclusion{Field: string(r.Status), Name: "Status", List: validStatuses},
+		&validators.StringInclusion{Field: string(r.MethodOfReceipt), Name: "Status", List: validMethodsOfReceipt},
```
Comment (akostibas) [inline pkg/models/reimbursement.go]: neat, didn't know about this one

File: pkg/models/reimbursement_test.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+	. "github.com/transcom/mymove/pkg/models"
+)
+
+func (suite *ModelSuite) TestReimbursementStateMachine() {
+	reimbursement := BuildDraftReimbursement(1200, MethodOfReceiptDIRECTDEPOSIT)
+
+	err := reimbursement.Request()
+	suite.Nil(err)
+	suite.Equal(reimbursement.Status, ReimbursementStatusREQUESTED, "expected Requested")
```
Comment (akostibas) [inline pkg/models/reimbursement_test.go]: Test `RequestedDate` got set to something around Now()?

File: pkg/models/reimbursement.go (hunk 2 above)
Comment (akostibas) [reply pkg/models/reimbursement.go]: Add a comment about how this might be used plz

File: pkg/models/reimbursement.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	ReimbursementStatusREQUESTED ReimbursementStatus = "REQUESTED"
+	// ReimbursementStatusAPPROVED captures enum value "APPROVED"
+	ReimbursementStatusAPPROVED ReimbursementStatus = "APPROVED"
+	// ReimbursementStatusREJECTED captures enum value "REJECTED"
+	ReimbursementStatusREJECTED ReimbursementStatus = "REJECTED"
+	// ReimbursementStatusPAID captures enum value "PAID"
+	ReimbursementStatusPAID ReimbursementStatus = "PAID"
+)
+
+// Reimbursement is money that is intedended to be paid to the servicemember
```
Comment (stangah) [inline pkg/models/reimbursement.go]: intended

File: pkg/models/reimbursement.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+	RequestedDate   *time.Time          `json:"requested_date" db:"requested_date"`
+}
+
+// State Machine
+// Avoid calling Reimbursement.Status = ... ever. Use these methods to change the state.
+
+// ErrInvalidTransition is an error representing an invalid transition.
+var ErrInvalidTransition = errors.New("INVALID_TRANSITION")
+
+// Request officailly requests the reimbursement.
```
Comment (stangah) [inline pkg/models/reimbursement.go]: officially

File: pkg/models/reimbursement.go [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+		return errors.Wrap(ErrInvalidTransition, "Approve")
+	}
+
+	r.Status = ReimbursementStatusAPPROVED
+	return nil
+}
+
+// Reject recjects the Reimbursement
+func (r *Reimbursement) Reject() error {
+	if r.Status != ReimbursementStatusREQUESTED {
```
Comment (stangah) [inline pkg/models/reimbursement.go]: Are there any valid transitions from REJECTED or APPROVED? Is the idea that an entirely new reimbursement model is requested is it needs to be reapproved or rescinded?

File: pkg/models/reimbursement.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+		return errors.Wrap(ErrInvalidTransition, "Reject")
+	}
+
+	r.Status = ReimbursementStatusREJECTED
+	return nil
+}
+
+// Pay pays the Reimbursement
+func (r *Reimbursement) Pay() error {
+	if r.Status != ReimbursementStatusAPPROVED && r.Status != ReimbursementStatusDRAFT {
```
Comment (stangah) [inline pkg/models/reimbursement.go]: DRAFT can be paid?

Comment (macrael) [reply pkg/models/reimbursement.go]: yeah, probably not.

File: pkg/models/reimbursement.go (hunk 7 above)
Comment (macrael) [reply pkg/models/reimbursement.go]: I don't know, I guess I made this pretty simple to start.

File: pkg/models/reimbursement.go (hunk 8 above)
Comment (macrael) [reply pkg/models/reimbursement.go]: Oh, I meant this to be SUBMITTED, but let's leave it out for now and add it if we need to.

File: pkg/models/reimbursement.go (hunk 7 above)
Comment (stangah) [reply pkg/models/reimbursement.go]: That's fair, we can add in complexity later

**Pull request merged**